### PR TITLE
Fix attribute terms disappearing in list

### PR DESCRIPTION
**Pull request by louwie17**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Small fix related to the changes of this previous PR: https://github.com/woocommerce/woocommerce/pull/39522

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Load this branch and enable the new product editor under **WooCommerce > Settings > Advanced > Features**
2. Install the WooCommerce Beta tester plugin and enable the `product-variation-management` feature under **Tools > WCA Test Helper > Features**
3. Go to **Products > Add New**
4. Add either some new global attributes or variation options (selecting global attributes)
5. Save the product
6. Refresh the page on the `Organization` tab or `Variations` tab
7. The terms should still be rendered within the list (previously they would disappear )

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/data/src/product-attribute-terms/types.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ type MutableProperties = Partial<
 
 type ProductAttributeTermActions = CrudActions<
 	'ProductAttributeTerm',
-	ProductAttributeTerm,
+	ProductAttributeTerm & { attribute_id: string },
```
Comment (louwie17) [inline packages/js/data/src/product-attribute-terms/types.ts:43]: Added as this is needed for the create attribute term request.

File: packages/js/data/src/product-attribute-terms/index.ts [hunk 2]
```diff
@@ -12,3 +20,14 @@ createCrudDataStore( {
...
+
+declare module '@wordpress/data' {
+	// TODO: convert action.js to TS
+	function dispatch( key: typeof STORE_NAME ): ActionDispatchers;
+	function select(
+		key: typeof STORE_NAME
+	): SelectFromMap< ProductAttributeTermsSelectors > & WPDataSelectors;
+	function resolveSelect(
+		key: typeof STORE_NAME
+	): PromiseifySelectors< SelectFromMap< ProductAttributeTermsSelectors > >;
```
Comment (louwie17) [inline packages/js/data/src/product-attribute-terms/index.ts:32]: This helps expose the store action/selector types automatically when using `useSelect` or `useDispatch`

File: packages/js/product-editor/src/hooks/use-product-attributes.ts [hunk 3]
```diff
@@ -69,12 +68,13 @@ export function useProductAttributes( {
...
-		terms: ProductAttributeTerm[]
+		allTerms: ProductAttributeTerm[]
 	) => {
 		return {
 			...globalAttribute,
-			terms: terms.length > 0 ? terms : undefined,
-			options: terms.length === 0 ? globalAttribute.options : [],
+			terms: ( allTerms || [] ).filter( ( term ) =>
+				globalAttribute.options.includes( term.name )
+			),
```
Comment (louwie17) [inline packages/js/product-editor/src/hooks/use-product-attributes.ts:77]: There was an issue here, where we always passed the `productId` to the `getProductAttributeTerms`, but this meant that any newly added global attributes would return an empty list as they were not saved yet.
I removed the `product` part so it always returns all the `terms` of the specific attribute. This did mean I had to add this `filter` to only add the terms that are part of the product, by comparing it to `options` (the list of term names ).

**Comment by louwie17** [comment]
@mdperez86 this should finally be ready for re-review :)

**Review by mdperez86 — APPROVED** [review]
LGTM @louwie17 nice job here!

**Pull request merged**